src/opt/pre_scheduling_impl.hpp

```diff
@@ -0,0 +1,204 @@
+#ifndef MIGRAPHX_GUARD_RTGLIB_PRE_SCHEDULING_IMPL_HPP
+#define MIGRAPHX_GUARD_RTGLIB_PRE_SCHEDULING_IMPL_HPP
+#include <migraphx/common_header.hpp>
+#include <migraphx/instruction_ref.hpp>
+#include <migraphx/operation.hpp>
+#include <migraphx/insert_instruction.hpp>
+
+namespace migraphx {
+
+struct dag_node
+{
+    dag_node()
+    {
+        weight         = 0;
+        run_on_cpu     = 0;
+        weight_sum     = 0;
+        ins_ndx        = -1;
+        first_child    = nullptr;
+        stream         = -1;
+        partition      = -1;
+        sched_cycle    = -1;
+        earliest_cycle = -1;
+    }
+    int weight;
+    int run_on_cpu;
+    int weight_sum;
+    int ins_ndx;
+    dag_node* first_child;
+    int stream;
+    int partition;
+    int sched_cycle;
+    int earliest_cycle = -1;
+    instruction_ref ins;
+    bool is_literal() const { return (ins->name() == "@literal"); }
+    bool can_use_stream() const { return (run_on_cpu == 0); }
```

Comment:
Actually, this can be written as `bool can_use_stream() const { return is_context_free(ins->get_operator()); }` since only operators that use a context can use a stream.